File: Solutions/Surface-Area-of-3D-Shapes.py
from typing import List

class Solution:
    def surfaceArea(self, grid: List[List[int]]) -> int:
        n = len(grid)
        ans = 0
        for i in range(n):
            for j in range(n):
                ans += grid[i][j] * 6
                if (grid[i][j] > 1): ans -= (grid[i][j] - 1) * 2
                if i > 0: ans -= min(grid[i - 1][j], grid[i][j])
                if i < n - 1: ans -= min(grid[i][j], grid[i + 1][j])
                if j > 0: ans -= min(grid[i][j - 1], grid[i][j])
                if j < n - 1: ans -= min(grid[i][j], grid[i][j + 1])
        return ans
    
S = Solution()
print(S.surfaceArea([[1,0],[0,2]]))

File: Solutions/Surface-Area-of-3D-Shapes.cpp
class Solution
{
  public:
    const vector<vector<int>> pos = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    int surfaceArea(vector<vector<int>> &grid)
    {
        int ans = 0, n = grid.size(), m = grid[0].size();
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                ans += grid[i][j] * 6 - 2 * max(0, grid[i][j] - 1);
                for (const vector<int> &p : pos)
                {
                    int x = i + p[0], y = j + p[1];
                    if (0 <= x && x < n && 0 <= y && y < m)
                        ans -= min(grid[i][j], grid[x][y]);
                };
            };
        };
        return ans;
    }
};